src/math_utils/convolution.py

```
@@ -0,0 +1,41 @@
+    kernel_height, kernel_width = kernel.shape
+    # Calculate padding dynamically based on kernel size
+    pad_h = kernel_height // 2
+    pad_w = kernel_width // 2
+    padded_image = np.pad(image_array, ((pad_h, pad_h), (pad_w, pad_w)), mode='constant')
+    
+    # Use scipy for efficient convolution
+import numpy as np
+from PIL import Image
+from scipy import signal
+
+def apply_convolution(image, kernel):
```

Ruff (0.8.2)

Redefinition of unused `apply_convolution` from line 4

(F811)